File: python/insertion_sort.py
def insertion_sort(array):
    for i in range(1, len(array)):
        currentElement = array[i]
        k = i - 1
        while k >= 0 and array[k] > currentElement:
            array[k + 1] = array[k]
            k -= 1
        array[k + 1] = currentElement


def insertion_sort_2(array):
    for i in range(1, len(array)):
        for k in range(i, 0, -1):
            if array[k - 1] > array[k]:
                temp = array[k - 1]
                array[k - 1] = array[k]
                array[k] = temp



print("testing insertion sort in python...")

unsortedArray = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray))
insertion_sort(unsortedArray)
print("sorted array -> " + str(unsortedArray))

print("testing insertion sort 2 in python...")

unsortedArray2 = [4, 1, 5, 2, 8, 3, 9, 7, 6]
print("unsorted array -> " + str(unsortedArray2))
insertion_sort_2(unsortedArray2)
print("sorted array -> " + str(unsortedArray2))

File: c/insertion_sort.c
#include <stdio.h>

void insertion_sort(int *array, int size);
void insertion_sort_2(int *array, int size);

int main() {
    int array[] = { 9, 5, 7, 1, 6, 2, 3, 8, 4 };
    int size = sizeof(array) / sizeof(int);

    insertion_sort(array, size);

    printf("testing insertion sort in c...\n");
    printf("sorted array -> [ %d", array[0]);
    for (int i = 1; i < size; i++) {
        printf(", %d", array[i]);
    }
    printf(" ]\n");

    return 0;
}

void insertion_sort(int *array, int size) {
    for (int i = 1; i < size; i++) {
        int currentElement = array[i];
        int k;
        for (k = i - 1; k >= 0 && array[k] > currentElement; k--) {
            array[k + 1] = array[k];
        }
        array[k + 1] = currentElement;
    }
}

void insertion_sort_2(int *array, int size) {
    for (int i = 1; i < size; i++) {
        for (int k = i; k > 0 && array[k - 1] > array[k]; k--) {
            int temp = array[k - 1];
            array[k - 1] = array[k];
            array[k] = temp;
        }
    }
}